pkg/storage/filepath/fs.go

```diff
@@ -186,23 +204,41 @@ func (fs *MemoryFS) Write(encoder runtime.Encoder, p string, obj runtime.Object,
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
 
-	versionKey := filepath.Clean(p)
-	if version, ok := fs.versions[versionKey]; ok && version != storageVersion {
+	if rawObj, err := fs.readBuffer(p); err != nil {
+		if os.IsNotExist(err) {
+			// storageVersion == 0 -> this is a create, so it's expected to not exist (continue)
+			// storageVersion != 0 -> object has been deleted, propagate err to avoid a zombie update
+			if storageVersion != 0 {
+				return err
+			}
+		} else {
+			return err
+		}
+	} else if rawObj.version != storageVersion {
+		// this write is outdated
 		return VersionError
+	} else if bytes.Equal(rawObj.data, buf.Bytes()) {
+		// object serialized identically, skip write & version increment
+		return nil
 	}
 
 	dir, err := fs.ensureDir(filepath.Dir(p))
 	if err != nil {
 		return err
 	}
 
-	newVersion, err := resourceVersion(obj)
-	if err != nil {
+	// increment the resource version - it's applied to the object pointer for
+	// the caller in addition to being used to ensure the write is valid
+	newVersion := atomic.AddUint64(&fs.rev, 1)
```

Comment:
i don't think you need an atomic here? you already have the Lock

Reply:
😅  removed in a3b384c